subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmFeature.java

```diff
@@ -248,31 +310,54 @@ public void withSourcesJar() {
         );
     }
 
-    private static Configuration bucket(String kind, RoleBasedConfigurationContainerInternal configurations, String configName, String displayName) {
-        Configuration configuration = configurations.maybeCreateWithRole(configName, ConfigurationRoles.INTENDED_BUCKET, false, false);
-        configuration.setDescription(kind + " dependencies for the " + displayName + ".");
+    private Configuration bucket(String kind, String suffix) {
+        String configName = getConfigurationName(suffix);
+        Configuration configuration = project.getConfigurations().maybeCreateWithRole(configName, ConfigurationRoles.INTENDED_BUCKET, false, false);
+        configuration.setDescription(kind + " dependencies for the '" + name + "' feature.");
         configuration.setVisible(false);
         return configuration;
     }
 
-    private TaskProvider<Task> registerOrGetJarTask(SourceSet sourceSet, TaskContainer tasks, String displayName) {
+    private TaskProvider<Jar> registerOrGetJarTask(SourceSet sourceSet, TaskContainer tasks) {
         String jarTaskName = sourceSet.getJarTaskName();
         if (!tasks.getNames().contains(jarTaskName)) {
-            tasks.register(jarTaskName, Jar.class, jar -> {
-                jar.setDescription("Assembles a jar archive containing the classes of the " + displayName + ".");
+            return tasks.register(jarTaskName, Jar.class, jar -> {
+                jar.setDescription("Assembles a jar archive containing the classes of the '" + name + "' feature.");
                 jar.setGroup(BasePlugin.BUILD_GROUP);
                 jar.from(sourceSet.getOutput());
-                jar.getArchiveClassifier().set(TextUtil.camelToKebabCase(name));
+                if (!capabilities.isEmpty()) {
+                    jar.getArchiveClassifier().set(TextUtil.camelToKebabCase(name));
+                }
             });
         }
-        return tasks.named(jarTaskName);
+        return tasks.named(jarTaskName, Jar.class);
     }
 
     @Override
     public CapabilitiesMetadata getCapabilities() {
         return ImmutableCapabilities.of(capabilities);
     }
 
+    @Override
+    public TaskProvider<Jar> getJarTask() {
+        return jar;
+    }
+
+    @Override
+    public TaskProvider<JavaCompile> getCompileJavaTask() {
+        return compileJava;
+    }
+
+    @Override
+    public SourceSetOutput getOutput() {
```

Comment:
This `getOutput()` method is a bit vague and since it just seems to duplicate `getSourceSet().getOutput()` I would vote for getting rid of it. Wdyt?

Reply:
This method was originally added when the intention was to phase out source sets in favor of interacting directly with this API -- this API exposes all the tasks and configurations which the source set only exposes the names of these domain objects.

Conceptually it makes sense that a source set should really only track a set of sources and maybe its dependencies. The fact that a source set also has outputs does not make sense in the context where a set of sources are compiled multiple times in different scenarios. 

However, we have de-prioritized work in this area so I am fine removing this until we have more progress in this area.